cpu/esp32/include/gpio_ll_arch.h

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2021 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     cpu_esp32
+ * @ingroup     drivers_periph_gpio_ll
+ * @{
+ *
+ * @file
+ * @brief       CPU specific part of the Peripheral GPIO Low-Level API
+ *
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @}
+ */
+
+#ifndef GPIO_LL_ARCH_H
+#define GPIO_LL_ARCH_H
+
+#include "gpio_arch.h"
+#include "soc/soc.h"
+#include "soc/gpio_struct.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifndef DOXYGEN /* hide implementation specific details from Doxygen */
+
+#define GPIO_PORT(num)          ((gpio_port_t)(&_esp32_ports[num]))
+#ifdef CPU_FAM_ESP32
+#define GPIO_PORT_NUM(port)     (((_esp32_port_t*)port == _esp32_ports) ? 0 : 1)
+#else
+#error "ESP32x family implementation missing"
+#endif
+
+/* GPIO port descriptor type */
+typedef struct {
+    volatile uint32_t *out;         /* address of GPIO_OUT/GPIO_OUT1 */
+    volatile uint32_t *out_w1ts;    /* address of GPIO_OUT_W1TS/GPIO_OUT1_W1TC */
+    volatile uint32_t *out_w1tc;    /* address of GPIO_OUT_W1TC/GPIO_OUT1_W1TC */
+    volatile uint32_t *in;          /* address of GPIO_IN/GPIO_IN1 */
+    volatile uint32_t *enable;      /* address of GPIO_ENABLE/GPIO_ENABLE1 */
+    volatile uint32_t *enable_w1ts; /* address of GPIO_ENABLE_W1TS/GPIO_ENABLE1_W1TS */
+    volatile uint32_t *enable_w1tc; /* address of GPIO_ENABLE_W1TC/GPIO_ENABLE1_W1TC */
+    volatile uint32_t *status_w1tc; /* address of GPIO_STATUS_W1TC/GPIO_STATUS1_W1TC */
+}  _esp32_port_t;
```

Comment:
Would it be possible to instead have a `struct` that matches the memory layout of the registers? That way one would only need to store the address of that `struct`, rather than the addresses of each register.

Reply:
There is already such structure defined in ESP-IDF. In that case port would be 0 or 1 and would have to be used as following:
```c
static inline void gpio_ll_set(gpio_port_t port, uword_t mask)
{
    if (port) {
        GPIO.out1_w1ts = mask;
    }
    else {
        GPIO.out1_w1ts.val = mask;
    }
}
```
I already had an implementation this way, but it was slower than using such a special structure with all addresses for each port and register. The latter costs only a few bytes in ROM. I think performance is more important than saving ROM on ESPs.

Reply:
Agreed :)